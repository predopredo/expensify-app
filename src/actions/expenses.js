// NPM MODULES
import uuid from 'uuid'
// FIREBASE DATABASE
import database from '../firebase/firebase';

// NO DB
// component calls action generator
// actions generator returns object
// component dispatches object
// redux store changes

// with DB
// component calls action generator
// actions generator returns function
// component dispatches function (?)
// function runs (has the ability to dispatch other actions and do whatever it wants)

// SET_EXPENSES
export const startSetExpenses = () => {
  return (dispatch) => {              //this function has access to dispatch (redux-thunk)
    // needs to add return so you can chain .then on app.js and render stuff
    return database.ref('expenses').once('value').then((snapshot) => {   //reads once from firebase and then...

      const expenses = [];
      snapshot.forEach((childSnapshot) => {            //populates the empty array
        expenses.push({
          id: childSnapshot.key,                       //sets firebase location (key) to id and saves it on redux store
          ...childSnapshot.val()                       //so we can use it for routing
        });

        dispatch(setExpenses(expenses));               //dispatches the populated array with mapped ids
      });
    });

  };
};

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// ADD_EXPENSE

// ***only works with redux-thunk wired up (on configureStore.js)***
export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData; // //created consts: description, note, amount and createdAt that have default values if no value is passed as argument
    const expense = { description, note, amount, createdAt }; // forms an object with above consts

    return database.ref('expenses').push(expense) // return for testing, won't be used in front-end
      .then((ref) => { // ref = newly added to database expense
        dispatch(addExpense({ // when new expense is written on db, you call dispatch and get id and everything else inside an object to store (and trigger rendering on connected components)
          id: ref.key,
          ...expense // you just get the id from the db, the rest is already on the front-end JS
        }));
      });
  };
};

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE', // Action type
  expense
});

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove()
      .then(() => {
        dispatch(removeExpense({ id }));
      })
  }
}

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update(updates)
      .then(() => {
        dispatch(editExpense(id, updates));
      })
  }
};