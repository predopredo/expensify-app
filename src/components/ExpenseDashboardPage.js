// Npm modules
import React from 'react';
// Components
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;