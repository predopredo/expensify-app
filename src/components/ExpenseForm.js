// NPM MODULES
import moment from 'moment';
import React from 'react';
// Airbnb date picker:
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';

// const now = moment();
// console.log(now.format('MMM Do, YYYY - HH:mm'));

export default class ExpenseForm extends React.Component {

  state = {
    description: this.props.expense ? this.props.expense.description : '',
    note: this.props.expense ? this.props.expense.note : '',
    amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
    createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
    calendarFocused: false,
    error: ''
  };

  onDescriptionChange = (event) => {
    const description = event.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = (event) => {
    const note = event.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = (event) => {
    const amount = event.target.value;
    // As it is a controlled input, the value will be always conditioned to the state so if you condition the user input to only get set as state if it matches a specific regex, anything not matching will simply be ignored and won't get into input value. That can be used as validation
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }))
    }
  };

  // airbnb date picker handlers:
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onDateFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  // Form Submission
  onSubmit = (event) => {
    event.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide description and amount!' }))
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmitToDispatch({
        description: this.state.description,
        amount: parseFloat((this.state.amount * 100).toFixed(0), 10),
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      })
    }

  };

  // Rendering JSX
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          autoFocus
          className="text-input"
          onChange={this.onDescriptionChange}
          placeholder="Description"
          type="text"
          value={this.state.description}
        />
        <input
          className="text-input"
          onChange={this.onAmountChange}
          placeholder="Amount"
          type="text"
          value={this.state.amount}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          displayFormat="DD/MM/YYYY"
          focused={this.state.calendarFocused}
          isOutsideRange={() => false}
          numberOfMonths={1}
          onDateChange={this.onDateChange}
          onFocusChange={this.onDateFocusChange}
        />
        <textarea
          className="textarea"
          onChange={this.onNoteChange}
          placeholder="Add a note for your expense (optional)"
          value={this.state.note}
        >
        </textarea>
        <div>
          <button className="button">{this.props.expense ? 'Update' : 'Add Expense'}</button>
        </div>
      </form>
    );
  };
};

