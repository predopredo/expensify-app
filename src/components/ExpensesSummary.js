// NPM MODULES
import React from 'react';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
// redux-react connect
import { connect } from 'react-redux';
// Selectors
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = (props) => {
  const expenseWord = props.expensesCount === 1 ? 'expense' : 'expenses';
  const formattedTotal = numeral(props.expensesTotal / 100).format('$0,0.00');
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Viewing <span>{props.expensesCount} </span>{expenseWord} totalling <span>{formattedTotal}</span>.
          <div className="page-header__actions">
            <Link className="button" to="/create">Add Expense</Link>
          </div>
        </h1>
      </div>
    </div>
  );
};

const mapStateToProps = (storedState) => {
  const visibleExpenses = selectExpenses(storedState.expenses, storedState.filters);
  return {
    expensesTotal: selectExpensesTotal(visibleExpenses),
    expensesCount: visibleExpenses.length
  };
};

export default connect(mapStateToProps)(ExpensesSummary); 