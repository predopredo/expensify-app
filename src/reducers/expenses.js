// Default state:
const expensesReducerDefaultState = [];

// Expenses Reducer
export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {

    case 'SET_EXPENSES':
      return action.expenses

    case 'ADD_EXPENSE':
      return [...state, action.expense];

    case 'REMOVE_EXPENSE':
      return state.filter((expense) => expense.id != action.id); // Only keeps items with id != from provided ((action.id) [when expense.id != action.id returns true]

    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates // get matching id expense and add each property from action.updates
          }
        } else {
          return expense; // if id doesn't match, just add the expense w/o modifying it
        }
      })

    default:
      return state;

  }
};
// ps: on REMOVE_EXPENSE: you can use state.filter(({ id }) => id != action.id) so filter({ id }) /=/ filter((expense) => expense.id ...) you can destructure a property of an object and pass it as the filter parameter to use it in filter condition