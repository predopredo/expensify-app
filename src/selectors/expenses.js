// NPM MODULES
import moment from 'moment';
// ***GET VISIBLE EXPENSES***

export default (expenses, { text, sortBy, startDate, endDate }) => {
  
  return expenses.filter((expense) => {

    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(moment(createdAtMoment, 'day')) : true; //moment compare functions
    const endDateMatch = endDate ? endDate.isSameOrAfter(moment(createdAtMoment, 'day')) : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
   
    return startDateMatch && endDateMatch && textMatch
     
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    }
    if (sortBy === 'amount') {
      return a.amount > b.amount ? -1 : 1;
    }
  });

};