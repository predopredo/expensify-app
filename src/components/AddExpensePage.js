// Npm modules
import React from 'react';
// Actions
import { startAddExpense } from '../actions/expenses'
// react-redux connect
import { connect } from 'react-redux';
// Components
import ExpenseForm from './ExpenseForm';

export class AddExpensePage extends React.Component {
  onSubmit = (formExpense) => { //*
    this.props.startAddExpense(formExpense); // ***
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container page-header--no-button ">
            <h1 className="page-header__title">Add Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm onSubmitToDispatch={this.onSubmit} //*
          />
        </div>

      </div>
    );
  }
};

// mapDispatchToProps let's you can return an object that somehow uses dispatch
// here we use it to abstract it away from component so this component gets easily testable with spy function
const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense)) // ***
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage);