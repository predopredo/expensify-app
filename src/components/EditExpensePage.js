// Npm modules
import React from 'react';
// react-redux
import { connect } from 'react-redux';
// actions
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
// Components
import ExpenseForm from './ExpenseForm';
import NotFoundPage from '../components/NotFoundPage'

export class EditExpensePage extends React.Component {
  render() {
    return this.props.expense ? (
      <div>
        <div className="page-header">
          <div className="content-container page-header--no-button ">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            expense={this.props.expense}
            onSubmitToDispatch={(FormExpense) => {
              this.props.startEditExpense(this.props.expense.id, FormExpense);
              this.props.history.push('/');
            }}
          />
          <button
            className="button button--secondary"
            onClick={() => {
              this.props.startRemoveExpense(this.props.expense.id);
              this.props.history.push('/');
            }}>Remove</button>
        </div>
      </div>
    ) :
      <div>
        <NotFoundPage />
      </div>
  }
}
// props = EditExpensesPage props
const mapStateToProps = (storedState, props) => ({
  expense: storedState.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (id) => dispatch(startRemoveExpense({ id }))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);