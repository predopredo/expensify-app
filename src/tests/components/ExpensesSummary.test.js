// NPM MODULES
import React from 'react';
import { shallow } from 'enzyme';
// COMPONENTS
import { ExpensesSummary } from '../../components/ExpensesSummary';
// SELECTORS
import selectExpensesTotal from '../../selectors/expenses-total';
// FIXTURES
import expenses from '../fixtures/expenses';


test('should render ExpensesSummary correctly with no expenses', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={0} expensesTotal={0}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary correctly with 1 expense', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={235}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary correctly with multiple expenses', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={23} expensesTotal={23512340987}/>);
  expect(wrapper).toMatchSnapshot();
});