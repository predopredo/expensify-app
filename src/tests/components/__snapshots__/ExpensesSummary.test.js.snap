// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary correctly with 1 expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
         
      </span>
      expense
       totalling 
      <span>
        $2.35
      </span>
      .
      <div
        className="page-header__actions"
      >
        <Link
          className="button"
          to="/create"
        >
          Add Expense
        </Link>
      </div>
    </h1>
  </div>
</div>
`;

exports[`should render ExpensesSummary correctly with multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        23
         
      </span>
      expenses
       totalling 
      <span>
        $235,123,409.87
      </span>
      .
      <div
        className="page-header__actions"
      >
        <Link
          className="button"
          to="/create"
        >
          Add Expense
        </Link>
      </div>
    </h1>
  </div>
</div>
`;

exports[`should render ExpensesSummary correctly with no expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        0
         
      </span>
      expenses
       totalling 
      <span>
        $0.00
      </span>
      .
      <div
        className="page-header__actions"
      >
        <Link
          className="button"
          to="/create"
        >
          Add Expense
        </Link>
      </div>
    </h1>
  </div>
</div>
`;
